rfcs/20240325-reorganize-gpu/README.md

```diff
@@ -0,0 +1,117 @@
+dependencies and have a flexible enough architecture to make adding support
+for new vendors easier.
+
+## Proposal
+
+Reorganize the GPU kernels and abstractions according the the following
+schema: Vendor / Technology
+
+With this schema the GPU directory will have subdirectories that
+correspond to the vendors: `intel`, `nvidia`, `amd`, `generic`, etc.
+Each of the subdirectories will have technology specific sub-subdirectories:
+`sycl`, `ocl`, `jit`, etc.
+
+Pros:
+* The schema provides enough flexibility to enable new vendor and extend
+the supported ones
+* Clustering functionality and abstractions around vendors is very
+convenient as a vendor can share the same functionality/abstractions across
+different technologies (e.g. the compute layer for Intel vendor)
+* Currently, functionality and abstractions are fully or partially
+clustered around vendors therefore the new schema should not cause a lot of
+confusion among the developers, it should also positively affect the
+implementation cost
+* The schema also provides sufficient configurability, e.g
+the generic vendor can be enabled along with nvidia or amd, or it can be
+enabled individually
+
+```bash
+├── cpu/ # CPU code
+├── sycl/ # A common place for basic CPU and GPU SYCL abstractions, e.g. `sycl_cpu_engine_t`, `sycl_gpu_engine_t`, etc.
+└── gpu/ # GPU code. Basic GPU code resides in the GPU directory directly, e.g. `gpu_engine_t`.
+    ├── intel/ # Everything that is related to Intel
+    │   ├── compute/ # Compute layer abstractions
+    │   ├── ocl/ # OpenCL kernels and abstractions (very similar to the current state of `gpu/ocl` directory)
+    │   ├── jit/ # JIT kernels/generators and abstractions (very similar to the current state of `gpu/jit` directory)
+    │   ├── sycl/ # Intel related SYCL functionality/abstractions (e.g. SYCL kernels that use eSIMD, Intel specific extensions, etc)
+    │   └── ...
+    ├── nvidia/ # Everything that is related to NVIDIA
+    │   ├── sycl # NVIDIA related functionality/abstractions (cuDNN/cuBLAS based and CUDA specific SYCL kernels)
+    │   └── ...
+    ├── amd/ # Everything that is related to AMD
+    │   ├── sycl/ # AMD related functionality/abstractions (MIOpen/rocBLAS based and HIP specific SYCL kernels)
+    │   └── ...
+    └── generic/ # Everything that is related to generic kernels
+        └── sycl/ # generic SYCL kernels and related abstractions
+
+```
+
+### Affected Basic Abstractions
+
+The new schema will require moving a lot of parts of the library around
+and while most of the changes are probably just an implementation detail
+there are a few major changes that have to be described in this RFC.
+
+#### Engine
+
+Engine is the most heavily affected abstraction by the changes.
+There is currently no `gpu_engine_t` class as `compute_engine_t` is used as
+an Intel centric alternative of it. In order to have a common basic
+abstraction at the GPU level the `gpu_engine_t` class will be introduced.
```

Comment:
I thought about that when I was writing this paragraph. I don't have a use case that would require `gpu_engine_t` class to exist at this point, it could be that there is none. The reason why I decided to add this class is to have something common at the GPU level. I'm currently working on an implementation and if I can't find any use cases I' may remove it as it won't be a big deal to add it when the need arises.

Reply:
Another idea that I'm mulling over is getting rid of `sycl_engine_base_t` (in the current RFC it's still there). The rational behind the idea is that I don't see how we can have a single version of `sycl_engine_base_t` given that it should be inherited from vendor specific classes (e.g. `compute_engine_t`) and I don't think that duplicating it is a good idea.

One option that I can see is to move the intel specific functionality over to `sycl_gpu_intel_engine_t`. After that, `sycl_engine_base_t` will just contain `sycl::context`, `sycl::device` and `backend_t`, and some basic interfaces such as `create_stream`, `create_memory_storage` and so on. With that in mind, I think we could go from inheritance to composition, for example,  `sycl_engine_base_t` can be renamed to `sycl_engine_impl_t` that would be included as data member into vendor specific classes such as `sycl_gpu_intel_engine_t`, `sycl_gpu_nvidia_engine_t` and `sycl_gpu_amd_engine_t`. The vendor specific classes would just need to provide interfaces that would redirect to the interfaces of `sycl_engine_impl_t`.

Any thoughts on that?

Reply:
Yeah, I like this idea, this should be cleaner than what we have now. Below is how it can be implemented (as I understand), this may or may not implementable but I think this should make things simpler:

- For Intel GPU engine we can have simple inheritance: `engine_t` -> `gpu_intel_engine_t`
    - `gpu_intel_engine_t` can include `shared_ptr<engine_impl_t>` and store `sycl_engine_impl_t` or `ocl_engine_impl_t`. Then `gpu_intel_engine_t` can do simple redirection via virtual methods of `engine_impl_t`
- `sycl_gpu_engine_t` can still be in the hierarchy for simplicity (as well as `sycl_cpu_engine_t`). It can store `sycl_engine_impl_t` and perform simple redirection. This way Nvidia/AMD engine classes can rely on simple inheritance and do not reimplement the same functionality
- `ocl_gpu_engine_t` is not really needed as all Intel-specific primitives will be implemented on top of `gpu_intel_engine_t`

Reply:
I updated the RFC, please take a look.